lib/peer-resolver/src/discovery.rs

```diff
@@ -0,0 +1,91 @@
+use std::time::Duration;
+
+use libp2p::{
+    futures::StreamExt,
+    identity::Keypair,
+    noise,
+    rendezvous::{client, Cookie},
+    swarm::{NetworkBehaviour, SwarmEvent},
+    tcp, yamux, Multiaddr, PeerId, Swarm, SwarmBuilder,
+};
+
+use crate::error::ResolverError;
+
+/// This struct holds the rendezvous behaviour for discovery.
+#[derive(NetworkBehaviour)]
+pub struct DiscoveryBehaviour {
+    rendezvous: client::Behaviour,
+}
+
+/// Rendezvous discover swarm
+pub struct DiscoverySwarm {
+    /// Swarm with [`DiscoveryBehaviour`]
+    swarm: Swarm<DiscoveryBehaviour>,
+    /// Rendezvous cookie for continuous peer discovery
+    cookie: Option<Cookie>,
+}
+
+impl DiscoverySwarm {
+    /// Create a new [`DiscoverySwarm`] with the given keypair.
+    /// The given timeout is set for the idle connection timeout
+    pub fn new<K>(keypair_bytes: K, timeout: u64) -> Result<DiscoverySwarm, ResolverError>
+    where
+        K: AsMut<[u8]>,
+    {
+        let keypair = Keypair::ed25519_from_bytes(keypair_bytes)?;
+        let swarm = SwarmBuilder::with_existing_identity(keypair)
+            .with_tokio()
+            .with_tcp(
+                tcp::Config::default(),
+                noise::Config::new,
+                yamux::Config::default,
+            )
+            .map_err(|_| ResolverError::InvalidTCPConfig)?
+            .with_behaviour(|key| DiscoveryBehaviour {
+                rendezvous: client::Behaviour::new(key.clone()),
+            })
+            .map_err(|_| ResolverError::InvalidBehaviourConfig)?
+            .with_swarm_config(|cfg| cfg.with_idle_connection_timeout(Duration::from_secs(timeout)))
+            .build();
+
+        Ok(DiscoverySwarm {
+            swarm,
+            cookie: None,
+        })
+    }
+
+    /// This functions requests the rendezvous_point for new or updated peers since the last cookie.
+    pub fn discover(&mut self, rendezvous_point: PeerId) {
```

Comment:
I'm not sure this function name is good, it isn't really descriptive of what it does. 
I know it wraps `discover` but I'm not convinced by libp2p's naming.